text/3201-return-type-attrs.md

```diff
@@ -0,0 +1,251 @@
+return type (because `rustfmt` by default puts each parameter on its own line
+when the function signature gets too long).
+
+Since function parameters already can be annotated this can be regarded as the
+next logical step towards more expressive and intuitive DSLs.  The motivation
+for the introduction of parameter attributes outlined in [RFC2565] largely
+applies to return type attributes as well, since they would also be useful for
+property based testing, interoperability with other languages and optimization
+annotations.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Return types of `fn` definitions as well as closures may have attributes
+attached to them. Thereby, additional information may be provided.
+
+For the purposes of illustration, let's assume we have the attribute
+`#[apple]` available to us.
+
+## Basic examples
+
+The syntax for attaching attributes to return types is shown in the snippet below:
+
+```rust
+// Free functions:
+fn foo() -> #[apple] u32 { .. }
+
+impl Alpha { // In inherent implementations.
+    fn bar() -> #[apple] u8 { .. }
+
+    ..
+}
+
+impl Beta for Alpha { // Also works in trait implementations.
+    fn bar() -> #[apple] u8 { .. }
+
+    ..
+}
+
+fn foo() {
+    // Closures:
+    let bar = |x| -> #[apple] u8 { .. };
+}
+```
+
+## Trait definitions
+
+An `fn` definition doesn't need to have a body to permit return type attributes.
+Thus, in `trait` definitions, we may write:
+
+```rust
+trait Beta {
+    fn bar(&self) -> #[apple] u8;
+}
+```
+
+## `fn` types
+
+You can also use attributes in function pointer types.
+For example, you may write:
+
+```rust
+type Foo = fn() -> #[apple] u8;
+```
+
+## Unit return type
+
+When annotating the unit return type `()` must be specified explicitly. For
+example:
+
+```rust
+fn foo() -> #[apple] () {
+    ..
+}
+```
+
+Attempting the following:
+
+```rust
+fn foo() -> #[apple] {
+    ..
+}
+```
+
+will result in a compile error:
+
+```
+error: return type attributes require an explicit return type
+fn foo() -> #[apple] {
+                    ^ expected ()
+```
+
+## Built-in attributes
+
+Attributes attached to return types do not have an inherent meaning in
+the type system or in the language. Instead, the meaning is what your
+procedural macros, the tools you use, or what the compiler interprets certain
+specific attributes as.
+
+As for the built-in attributes and their semantics, we will, for the time being,
+only permit the following attributes on return types:
+
+- Lint check attributes, that is:
+  `#[allow(C)]`, `#[warn(C)]`, `#[deny(C)]`, `#[forbid(C)]`,
+  and tool lint attributes such as `#[allow(clippy::foobar)]`.
+
+All other built-in attributes will be rejected with a semantic check.
+For example, you may not write:
+
+```rust
+fn foo() -> #[inline] u32 { .. }
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+TODO
```

Comment:
It would be good to check the parser's grammar to ensure that adding an attribute before the return type does not break anything (I don't expect that it will).